convert2rhel/applock.py

```diff
@@ -0,0 +1,186 @@
+    PID and checks to see if the process is still running. If the
+    process is still running, it raises ApplicationLockedError. If the
+    process is not running, it removes the file and tries to lock it
+    again.
+
+    For safety, unexpected conditions, like garbage in the file or
+    bad permissions, are treated as if the application is locked,
+    because something is obviously wrong.
+    """
+
+    def __init__(self, name):
+        # Our application name
+        self._name = name
+        # Do we think we locked the pid file?
+        self._locked = False
+        # Maximum number of tries to lock
+        self._max_loop_count = 5
+        # Our process ID
+        self._pid = os.getpid()
+        # Path to the file that contains the process id
+        self._pidfile = os.path.join(_DEFAULT_LOCK_DIR, self._name + ".pid")
+
+    def __str__(self):
+        if self._locked:
+            status = "locked"
+        else:
+            status = "unlocked"
+        return "%s PID %d %s" % (self._pidfile, self._pid, status)
+
+    def _try_create(self):
+        """Try to create the lock file. If this succeeds, the lock file
+        exists and we created it. If an exception other than the one
+        we expect is raised, re-raises it.
+
+        :returns: True if we created the lock, False if we didn't.
+        """
+        #
+        # Create a temporary file that contains our PID and attempt
+        # to link it to the real pidfile location. The link will fail if
+        # the file already exists; this avoids a race condition when
+        # two processes attempt to create the file simultaneously. This
+        # also guarantees that the lock file contains valid data.
+        #
+        with tempfile.NamedTemporaryFile(mode="w", suffix=".pid", prefix=self._name, dir=_DEFAULT_LOCK_DIR) as f:
+            f.write(str(self._pid) + "\n")
+            f.flush()
+            try:
+                os.link(f.name, self._pidfile)
+            # In Python 3 this could be changed to catch FileExistsError.
+            except OSError as exc:
+                if exc.errno == errno.EEXIST:
+                    return False
+                raise exc
+        loggerinst.debug("%s." % self)
+        return True
+
+    @property
+    def is_locked(self):
+        """Test whether this object is locked."""
+        return self._locked
+
+    @staticmethod
+    def _pid_exists(pid):
+        """Test whether a particular process exists."""
+        try:
+            # Bulletproofing: avoid killing init or all processes.
+            if pid > 1:
+                os.kill(pid, 0)
+        except OSError as exc:
+            #
+            # The only other (theoretical) possibility is EPERM, which
+            # would mean the process exists.
+            #
+            if exc.errno == errno.ESRCH:
+                return False
+        return True
+
+    def try_to_lock(self, loop_count=0):
```

Comment:
The only purpose for the `loop_count` variable is to make sure some strange set of circumstances can't throw us into an infinite loop. Toshio suggested removing the is_locked method and the associated field, which I will do in a post-merge-freeze update. As for supporting multiple locks/unlocks within the same process, I didn't do that because it was never presented as a requirement (and nobody hinted about a need for it). It would be easy enough to add a refcount, but if nobody needs it I'd rather not write it in the first place. If we do need it in the future, is it because we've started running multiple asynchronous threads? Will "refcount += 1" work?

Reply:
Over the weekend I've been thinking about the need for locks if we introduce parallelism in the code.  If we do that, I think we'll want to use a different sort of lock (multiprocessing and threading both have their own `Lock` structures for instance).

I don't think we'll want to release the application-wide lock while conver2rhel is running.  It's really only safe to unlock before we get to the `pre_ponr_changes` Actions.  And even though it's safe before then, it doesn't add value because it means a second convert2rhel started at the right time will start to do work that doesn't modify the system but, by the time it is ready to acquire the lock, it will fail because the first one has now locked the system.